c_src/libSigPyProc.cpp

```diff
@@ -0,0 +1,674 @@
+            if (indata[(ii * nchans) + jj] > flagMax_arr[jj])
+                flag_arr[(ii * nchans) + jj] = 2;
+            else if (indata[(ii * nchans) + jj] < flagMin_arr[jj])
+                flag_arr[(ii * nchans) + jj] = 0;
+            else
+                flag_arr[(ii * nchans) + jj] = 1;
+        }
+    }
+}
+
+
+/**
+ * TODO: This code is very slow compared to python slicing.
+ */
+template <class T>
+void splitToBands(py::array_t<T> inarray, py::array_t<T> outarray, int nchans,
+                  int nsamps, int nsub, int chanpersub, int chanstart) {
+    py::buffer_info inbuf = inarray.request(), outbuf = outarray.request();
+
+    T* indata  = (T*)inbuf.ptr;
+    T* outdata = (T*)outbuf.ptr;
+
+    int chan_end = nsub * chanpersub + chanstart;
+#pragma omp parallel for default(shared)
+    for (int ii = 0; ii < nsamps; ii++) {
+        for (int jj = chanstart; jj < chan_end; jj++) {
+            outdata[(ii * nchans) +
+                    (jj % chanpersub * nsub + jj / chanpersub)] =
+                indata[(ii * nchans) + jj];
+        }
+    }
+}
+
+template <class T>
+void getTim(py::array_t<T> inarray, py::array_t<float> outarray, int nchans,
+            int nsamps, int index) {
+    py::buffer_info inbuf = inarray.request(), outbuf = outarray.request();
+
+    T*     indata  = (T*)inbuf.ptr;
+    float* outdata = (float*)outbuf.ptr;
+#pragma omp parallel for default(shared)
+    for (int ii = 0; ii < nsamps; ii++) {
+        for (int jj = 0; jj < nchans; jj++) {
+            outdata[index + ii] += indata[(nchans * ii) + jj];
+        }
+    }
+}
+
+template <class T>
+void getBpass(py::array_t<T> inarray, py::array_t<double> outarray, int nchans,
+              int nsamps) {
+    py::buffer_info inbuf = inarray.request(), outbuf = outarray.request();
+
+    T*      indata  = (T*)inbuf.ptr;
+    double* outdata = (double*)outbuf.ptr;
+#pragma omp parallel for default(shared)
+    for (int jj = 0; jj < nchans; jj++) {
+        for (int ii = 0; ii < nsamps; ii++) {
+            outdata[jj] += indata[(nchans * ii) + jj];
+        }
+    }
+}
+
+template <class T>
+void dedisperse(py::array_t<T> inarray, py::array_t<float> outarray,
+                py::array_t<int32_t> delays, int maxdelay, int nchans,
+                int nsamps, int index) {
+    py::buffer_info inbuf = inarray.request(), outbuf = outarray.request();
+    py::buffer_info delaysbuf = delays.request();
+
+    T*       indata     = (T*)inbuf.ptr;
+    float*   outdata    = (float*)outbuf.ptr;
+    int32_t* delays_arr = (int32_t*)delaysbuf.ptr;
+
+#pragma omp parallel for default(shared)
+    for (int ii = 0; ii < (nsamps - maxdelay); ii++) {
+        for (int jj = 0; jj < nchans; jj++) {
+            outdata[index + ii] +=
+                indata[(ii * nchans) + (delays_arr[jj] * nchans) + jj];
+        }
+    }
+}
+
+template <class T>
+void maskChannels(py::array_t<T> inarray, py::array_t<uint8_t> mask, int nchans,
```

Comment:
question: why is mask 8-bit, not boolean?

Reply:
It was 8-bit in the original code. boolean will be more appropriate.